### PR TITLE
VACMS-6057 recurring events instances modal logic

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
#6057

## Testing done
Visual / behat

## Screenshots
image

## QA steps
 - Will be collaborative / in real time with Rachel and Kev

### CODE REVIEW

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+   *
+   * @return array
+   *   A render array structure of fields for this entity.
+   *
+   * @see \Drupal\Core\Entity\EntityListBuilder::render()
+   */
+  public function buildRow(array $instance) {
+    // Get format settings.
+    // @todo make the choice of format configurable?
+    $format = \Drupal::getContainer()
```
Comment (ethanteague) [inline docroot/modules/custom/va_gov_events/src/Controller/Instances.php:196]: The files in this dir are copypastas from smart_date_recur, which doesn't use DI. I changed logic as necessary to point to changes to text, etc. in va_gov_events, and applied phpcs changes. We could use DI, but imo, this will lead to potential headaches in the future if/when we need to update these files to track changes made to smart_date_recur.
@swirtSJW et.al.^^^^^^^

**Comment by kevwalsh** [comment]
We should meet to go over this later, but this screenshot captures most of my concerns


**Placement of Manage event series and Time zone in vagovadmin (Seven) theme**

<details><summary>Screenshot</summary>
Edit_Event_asdf___VA_gov_CMS



</details>

1. Putting "Edit event series" (and Time Zone for that matter) way over to the right is not a common pattern for fields or buttons like this. I know i'm late in the game to say, but seeing it in action on my big monitor (or even at a common size for our editors, 20-24" makes it them harder to see)
2. At the very least, the modal loading icon needs to be connected to the button.
Suggestion: we solve these issues with Claro-specific theming (see below) 


**Unchecked "Make recurring" checkbox is confusing for existing events that are already reccurring**
1. Creating an event that's every week
2. Go back to edit, 
3. See that the Make recurring is unchecked. (which is on its own a bit confusing)
4. Check it again, and make it recur every day
5. Save
6. See that the changes you made to make it recur every day didn't do anything

Suggestion: we either
1. Fix it so that changes to recurring event patterns that are made to existing recurring events actually work, and the checkbox would be checked by default. (idealy but probably difficult)
2. Just hide the Make recurring checkbox for existing events that are already recurring. 
If we go with 2, i wonder if it makes sense to float the "Manage event series" back to the left. 

**Claro**
I just enabled Claro on [this tugboat environment](https://pr7713-faszuoll2jsttdybs8hfteejfge0kcfp.ci.cms.va.gov/node/43232/edit) to see what would happen. Things are kind of broken, but looks like minor fixes
* Start and end field labels are overlapping with their fields
* when selecting weekly pattern, "on these days" fields  are overlapping with the other recurring fields
* when selecting monthly pattern, "on the" fields fields are overlapping with the other recurring fields
* Loading icon is not attached to Manage series button
* "Show row weights" is visible but doesn't do anything because we only allow one pattern, and should be hidden
* Make recurring checkbox has a bad margin


<details><summary>Screenshots</summary>
Edit_Event_asdf___VA_gov_CMS

Edit_Event_asdf___VA_gov_CMS


</details>

**Comment by rachel-kauff** [comment]
> Putting "Edit event series" (and Time Zone for that matter) way over to the right is not a common pattern for fields or buttons like this. I know i'm late in the game to say, but seeing it in action on my big monitor (or even at a common size for our editors, 20-24" makes it them harder to see)

I hadn't considered the screen size issue this might cause, and it certainly makes sense to move those righthand fields back for now, but this does make me think about how to think about the width of forms in future iterations of claro/design sys. I don't think the width should scale infinitely.

**Comment by rachel-kauff** [comment]
Screen Shot 2022-01-26 at 7 53 18 AM

**Comment by rachel-kauff** [comment]
Re: "Make recurring checkbox is unchecked when you go back to edit a recurring event."
I want to advocate hard for this one:

> Fix it so that changes to recurring event patterns that are made to existing recurring events actually work, and the checkbox would be checked by default. (idealy but probably difficult)

It's really confusing to not be able to see the changes I just made. What if I want to adjust what I just did to make my event repeat every two weeks instead of every one? The interface I had oriented myself to suddenly is gone and I start to lose trust that the content I had previously saved actually worked.

Screen Shot 2022-01-26 at 8 04 15 AM

**Comment by rachel-kauff** [comment]
A few language changes to the buttons in this modal (What's in the PR still uses some of the OOB labels, but I also had some second thoughts about what was in the Figma.) 

Here's what's there:
Screen Shot 2022-01-26 at 8 43 09 AM

I think these button labels clarify a little more:
Screen Shot 2022-01-26 at 8 43 23 AM

**Comment by ethanteague** [comment]
@kevwalsh @rachel-kauff will need to get with @tjheffner & @swirtSJW for an assist on sorting out the behavioral failures (there's some environment specific stuff that needs to be sorted out / doesn't occur in local instance), but I've applied the latest version of smart_date and made recurring logic & the rest of the events page compatible with Claro.

image

image

File: config/sync/system.theme.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-admin: vagovadmin
-default: vagovadmin
+admin: vagovclaro
+default: vagovclaro
```
Comment (tjheffner) [inline config/sync/system.theme.yml:2]: we don't want to commit this change yet. it'll turn on claro theme for everyone

Comment (ethanteague) [reply config/sync/system.theme.yml:2]: Yep, with you there, @tjheffner - was thinking this wouldn't be merged until claro was released.

Comment (ethanteague) [reply config/sync/system.theme.yml]: turned off in pr

File: tests/behat/features/save_and_continue.feature [hunk 3]
```diff
@@ -44,8 +44,8 @@ Feature: Save and continue button works as expected.
 
     # Confirm that the moderation history and state are shown correctly.
     And I should see "Draft" in the ".views-field-moderation-state" element
-    Then I visit the "moderation-history" page for a node with the title "Behat save and continue new test - edited"
-    And I should see "Set to Draft on " in the ".views-field-moderation-state" element
```
Comment (swirtSJW) [inline tests/behat/features/save_and_continue.feature:48]: Why is this changing with events?

File: tests/behat/features/user.feature [hunk 4]
```diff
@@ -5,7 +5,6 @@ Feature: The VA Website is accessible to users.
     Given I am on the homepage
     Then I should see "Log in"
     And I should see "Veterans Affairs" in the "title" element
-    And I should see "Login with PIV or other Smartcard." in the "#edit-simplesamlphp-auth-login-link" element
```
Comment (swirtSJW) [inline tests/behat/features/user.feature:8]: events?

File: tests/behat/features/content_editing.feature [hunk 5]
```diff
@@ -72,8 +72,8 @@ Feature: CMS Users may effectively create & edit content
     And I press "Save"
 
     # Confirm that the va.gov url is not shown for nodes without a published revision.
-    Then I should see "Content Type: Office" in the "#block-entitymetadisplay" element
-    And I should not see "VA.gov URL" in the "#block-entitymetadisplay" element
+    Then I should see "Content Type: Office" in the "#block-vagovclaro-entitymetadisplay" element
```
Comment (swirtSJW) [inline tests/behat/features/content_editing.feature:75]: ALl these tests changes seem unrelated to evetns.

**Review by swirtSJW — COMMENTED** [review]
Seeing some changes that don't seem related to events

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 6]
```diff
@@ -74,6 +74,16 @@ public function __construct(
...
   }
 
+  /**
+   * Alteration to entity view pages.
+   *
+   * @param \Drupal\core_event_dispatcher\Event\Entity\EntityViewAlterEvent $event
+   *   The entity view alter service.
+   */
+  public function entityViewAlter(EntityViewAlterEvent $event):void {
+    $this->appendHealthServiceTermDescriptionToVetCenter($event);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:84]: Is this related?   It seems like this whole file should be removed from this PR.   All the changes seem odd... stuff is jumping around and unrelated.

File: config/sync/field.field.node.event.field_address.yml [hunk 7]
```diff
@@ -44,11 +44,19 @@ settings:
...
+      override: optional
+    addressLine2:
+      override: optional
     postalCode:
       override: hidden
     sortingCode:
       override: hidden
     dependentLocality:
       override: hidden
+    locality:
```
Comment (swirtSJW) [inline config/sync/field.field.node.event.field_address.yml:57]: related?

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 8]
```diff
@@ -218,7 +232,7 @@ public function getProductTypeTermFields(array &$form, FormStateInterface $form_
   public function appendHealthServiceTermDescriptionToVetCenter(EntityViewAlterEvent $event):void {
     if ($event->getDisplay()->getTargetBundle() === 'vet_center') {
       $build = &$event->getBuild();
-      $services = isset($build['field_health_services']) ? $build['field_health_services'] : [];
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:221]: related?

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+  protected $formBuilder;
+
+  /**
+   * Indicating whether current controller instance uses Ajax.
+   *
+   * @var bool
+   */
+  private $useAjax;
+
+  /**
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_events/src/Controller/Instances.php:68]: can't inherit when you override with create()

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+        }
+        $instance['class'] = $override_type;
+        $instance['override'] = $override;
+      }
+      else {
+      }
+      $instance['rrule'] = $this->rrule->id();
+      $instance['rrule_index'] = $index;
+
+    }
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_events/src/Controller/Instances.php:148]: need an unset here to break the reference on $instance

**Review by swirtSJW — COMMENTED** [review]
Ok This should be the last round of review.   I found a couple things.... I think

File: config/sync/views.view.non_clinical_services.yml [hunk 11]
```diff
@@ -1124,7 +1124,7 @@ display:
           empty: true
           tokenize: false
           content:
-            value: 'You have not added a register for care location yet. '
```
Comment (swirtSJW) [inline config/sync/views.view.non_clinical_services.yml:1127]: Should this file be in this pr

File: docroot/modules/custom/va_gov_events/va_gov_events.services.yml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+services:
+  va_gov_events.entity_event_subscriber:
+    class: Drupal\va_gov_events\EventSubscriber\EntityEventSubscriber
+    arguments: ['@string_translation']
+    tags:
+      - { name: event_subscriber }
+  va_gov_events.route_subscriber:
+    class: Drupal\va_gov_events\Routing\RouteSubscriber
+    tags:
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_events/va_gov_events.services.yml:9]: The arguments that you were loading with create... should actually be defined here.  Then the create is not needed.

File: config/sync/field.field.node.event.field_address.yml (hunk 7 above)
Comment (ethanteague) [reply config/sync/field.field.node.event.field_address.yml:57]: Yep, @swirtSJW  - this is events form logic from https://github.com/department-of-veterans-affairs/va.gov-cms/issues/2460 that we bundled with recurring.

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php (hunk 6 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:84]: Agreed - removed, @swirtSJW

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php (hunk 8 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:221]: removed

Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:221]: Actually, we need this b/c PHPstan was complaining:  `Use null coalesce operator instead of ternary operator.`

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php (hunk 9 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_events/src/Controller/Instances.php:68]: Don't understand what you mean here, @swirtSJW

File: config/sync/views.view.non_clinical_services.yml (hunk 11 above)
Comment (ethanteague) [reply config/sync/views.view.non_clinical_services.yml:1127]: Removed

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php (hunk 10 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_events/src/Controller/Instances.php:153]: @swirtSJW straightened me out in slack

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php (hunk 9 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_events/src/Controller/Instances.php:68]: Was reading the wrong line - resolved.

File: tasks-periodic.yml [hunk 13]
```diff
@@ -94,7 +94,7 @@ tasks:
   va/background/daily/migrate/flag_removed_facilities:
     desc: These flagging process is to run once nightly.
     cmds:
-      - drush $DRUSH_ALIAS drush va_gov_migrate:flag-missing-facilities
+      - drush $DRUSH_ALIAS va_gov_migrate:flag-missing-facilities
```
Comment (swirtSJW) [inline tasks-periodic.yml:97]: merge bubble  This should not be here

File: docroot/modules/custom/va_gov_events/src/Controller/Instances.php [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+        }
+        $instance['class'] = $override_type;
+        $instance['override'] = $override;
+      }
+      else {
+      }
+      $instance['rrule'] = $this->rrule->id();
+      $instance['rrule_index'] = $index;
+      unset($instance);
+    }
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_events/src/Controller/Instances.php:153]:
```suggestion
    }
    https://github.com/department-of-veterans-affairs/va.gov-cms/pull/7713
```

**Review by swirtSJW — APPROVED** [review]
Herculean effort brother.  Well done.

File: tests/behat/features/content_editing.feature [hunk 15]
```diff
@@ -83,31 +84,32 @@ Feature: CMS Users may effectively create & edit content
...
     Then I visit the "edit" page for a node with the title "Test Office - BeHaT 404"
     And I select "Archived" from "edit-moderation-state-0-state"
     And I fill in "Revision log message" with "Test archiving"
     And I press "Save"
-    Then I should see "Test Office - BeHaT 404\u003C\/a\u003E\u003C\/em\u003E has been updated."
-    And I should see "Content Type: Office" in the "#block-entitymetadisplay" element
-    And I should not see "VA.gov URL" in the "#block-entitymetadisplay" element
+    Then I should see "has been updated."
+    And I should see "Content Type: Office" in the "#block-vagovclaro-entitymetadisplay" element
+    And I should not see "VA.gov URL" in the "#block-vagovclaro-entitymetadisplay" element
```
Comment (tjheffner) [inline tests/behat/features/content_editing.feature:112]: so i was fixing this in https://github.com/department-of-veterans-affairs/va.gov-cms/pull/7953 in a way that works in both themes... i'm worried this will fail tests regardless of which theme is built now
